orderer/consensus/hlmirbft/chain.go

```diff
@@ -1502,3 +1370,40 @@ func (c *Chain) triggerCatchup(sn *raftpb.Snapshot) {
 	case <-c.doneC:
 	}
 }
+
+// TODO(harrymknight) Implement these methods
+func (c *Chain) Apply(*msgs.QEntry) error {
+	return nil
+}
+
+//FLY2-66 JIRA issue code change
+func (c *Chain) Snap(networkConfig *msgs.NetworkState_Config, clientsState []*msgs.NetworkState_Client) ([]byte, []*msgs.Reconfiguration, error) {
+	var persistBytes []byte
+	pr := c.Node.PendingReconfigurations
+	c.Node.PendingReconfigurations = nil
+	c.Node.CheckpointSeqNo = c.Node.LastSeqNo
+
+	c.Node.CheckpointState = &msgs.NetworkState{
+		Config:                  networkConfig,
+		Clients:                 clientsState,
+		PendingReconfigurations: pr,
+	}
+	c.Node.CheckpointHash = c.Node.ActiveHash.Sum(nil)
+	c.Node.ActiveHash = c.Node.Hasher.New()
+	c.Node.ActiveHash.Write(c.Node.CheckpointHash)
+	checkpointBytes, err := proto.Marshal(c.Node.CheckpointState)
+	if err != nil {
+		return nil, nil, err
+	}
+	persistBytes = append([]byte{}, c.Node.CheckpointHash...)
+	persistBytes = append(persistBytes, checkpointBytes...)
+	err = c.Node.PersistSnapshot(c.appliedIndex, persistBytes)
```

Comment:
c.appliedIndex is currently tied to an index which is incremented when a block is written to the ledger. We will have multiple leaders and therefore will receive batches out of order (seqNo are distributed across leaders using modulo arithmetic) meaning we won't know what seqNo this index will respond. Given that a snapshot is to be retrieved via the TransferTo method when given its persisted value and a seqNo it may be better to persist a snapshot using a seqNo instead of c.index. I'll make a note find out why a seqNo is used to call TransferTo.

Reply:
The mentioned "seqNo" is the new field in the Chain struct just to keep the track of snap shot index ?
Or Is it par of the Mir BFT lib?